src/meta-srv/src/service/heartbeat.rs

```diff
@@ -136,6 +155,27 @@ async fn handle_ask_leader(req: AskLeaderRequest, ctx: Context) -> Result<AskLea
     Ok(AskLeaderResponse { header, leader })
 }
 
+fn get_node_id(req: &HeartbeatRequest) -> Result<u64> {
+    static ID: OnceCell<Arc<AtomicU64>> = OnceCell::new();
+
+    fn next_id() -> u64 {
+        let id = ID.get_or_init(|| Arc::new(AtomicU64::new(0))).clone();
+        id.fetch_add(1, std::sync::atomic::Ordering::Relaxed)
+    }
+
+    let header = req
+        .header
+        .as_ref()
+        .context(error::MissingRequestHeaderSnafu)?;
+
+    let node_id = match header.role() {
+        Role::Frontend => next_id(),
+        Role::Datanode => req.peer.as_ref().map(|p| p.id).unwrap_or(header.member_id),
```

Comment:
I'm a little confused here... Why `p.id` and `header.member_id` can both represent Datanode's id? What if the two diverge? I think we can just keep `header.member_id` and get rid of `p.id`. WDYT? @fengjiachun

Reply:
<header.member_id> represents the ID of the node sending the heartbeat, while <peer.id> represents the ID of the node corresponding to <peer.addr> In the context we are currently discussing, they are actually equal. However, in other scenarios such as [Replica.peer,](https://github.com/GreptimeTeam/greptime-proto/blob/main/proto/greptime/v1/meta/heartbeat.proto#L76) it may not be equal to <header.member_id>.

Above, I was just explaining their respective meanings. Now let's get back to our question. Thank you for your suggestion. I think using <header.member_id> here is better. How about to change it to header.member_id? @Fengys123

Reply:
Okay, I will modify it.